usage/src/org/apache/cloudstack/quota/QuotaManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+        if (quotaListForAccount == null || quotaListForAccount.size() < 1) {
+            return;
+        }
+        quotaListForAccount.add(new QuotaUsageVO());
+        Date startDate = quotaListForAccount.get(0).getStartDate();
+        Date endDate = quotaListForAccount.get(0).getEndDate();
+        BigDecimal aggrUsage = new BigDecimal(0);
+        for (QuotaUsageVO entry: quotaListForAccount) {
+            if (startDate.compareTo(entry.getStartDate()) != 0) {
+                QuotaBalanceVO lastRealBalanceEntry = _quotaBalanceDao.findLastBalanceEntry(account.getAccountId(), account.getDomainId(), startDate);
+                Date lastBalanceDate = new Date(0);
+                if (lastRealBalanceEntry != null) {
+                    lastBalanceDate = lastRealBalanceEntry.getUpdatedOn();
+                    aggrUsage = aggrUsage.add(lastRealBalanceEntry.getCreditBalance());
+                }
+
+                List<QuotaBalanceVO> creditsReceived = _quotaBalanceDao.findCreditBalance(account.getAccountId(), account.getDomainId(), lastBalanceDate, endDate);
+                if (creditsReceived != null) {
+                    for (QuotaBalanceVO credit : creditsReceived) {
+                        aggrUsage = aggrUsage.add(credit.getCreditBalance());
+                    }
+                }
+
+                QuotaBalanceVO newBalance = new QuotaBalanceVO(account.getAccountId(), account.getDomainId(), aggrUsage, endDate);
+
+                if (s_logger.isDebugEnabled()) {
+                    s_logger.debug("Balance entry=" + aggrUsage + " on Date=" + endDate);
+                }
+                _quotaBalanceDao.persist(newBalance);
+                aggrUsage = new BigDecimal(0);
+            }
+            startDate = entry.getStartDate();
+            endDate = entry.getEndDate();
+            aggrUsage = aggrUsage.subtract(entry.getQuotaUsed());
+        }
+
+        // update quota_accounts
+        QuotaAccountVO quota_account = _quotaAcc.findById(account.getAccountId());
+        if (s_logger.isDebugEnabled()) {
+            s_logger.debug("Updating quota account bal=" + aggrUsage + " date=" + endDate);
+        }
+        if (quota_account == null) {
+            quota_account = new QuotaAccountVO(account.getAccountId());
+            quota_account.setQuotaBalance(aggrUsage);
+            quota_account.setQuotaBalanceDate(endDate);
+            _quotaAcc.persist(quota_account);
+        } else {
+            quota_account.setQuotaBalance(aggrUsage);
+            quota_account.setQuotaBalanceDate(endDate);
+            _quotaAcc.update(account.getAccountId(), quota_account);
+        }
+    }
+
+    @Override
+    public boolean calculateQuotaUsage() {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        boolean jobResult = false;
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
```

Comment:
On second look, we are using a cloud schema object to query usage db. It is kind of convoluted and finds it roots in the current usage code.
This should be cleaned up, but it is not straight forward as just putting the code in a DAO.